File: go/0518_coin_change_ii.go
func change(amount int, coins []int) int {
	if amount == 0 {
		return 1
	} else if len(coins) == 0 {
		return 0
	} else if len(coins) == 1 {
		if amount%coins[0] == 0 {
			return 1
		}
		return 0
	}

	visited := make(map[Key]int)
	return doChange(amount, coins, visited)
}

func doChange(amount int, coins []int, visited map[Key]int) int {
	clen := len(coins)

	if amount == 0 {
		return 1
	} else if clen == 0 {
		return 0
	} else if clen == 1 {
		if amount%coins[0] == 0 {
			return 1
		}
		return 0
	}

	if ways, ok := visited[Key{Amount: amount, CoinLen: clen}]; ok {
		return ways
	}

	coin := coins[0]
	ways := 0
	for i := 0; i*coin <= amount; i++ {
		ways += doChange(amount-i*coin, coins[1:], visited)
	}

	visited[Key{Amount: amount, CoinLen: clen}] = ways
	return ways
}

// Key for visited map
type Key struct {
	Amount  int
	CoinLen int
}

File: python/0518_coin_change_ii.py
class Solution(object):
    def change(self, amount, coins):
        """
        :type amount: int
        :type coins: List[int]
        :rtype: int
        """

        dp = [0] * (amount + 1)
        dp[0] = 1
        for coin in coins:
            for i in range(1, amount + 1):
                if i - coin >= 0:
                    dp[i] += dp[i - coin]
            print dp
        return dp[amount]
